components/supervisor/pkg/ports/ports-config.go

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) 2020 TypeFox GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package ports
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+
+	"github.com/gitpod-io/gitpod/supervisor/pkg/gitpod"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+)
+
+// RangeConfig is a port range config
+type RangeConfig struct {
+	*gitpod.PortsItems
+	Start uint32
+	End   uint32
+}
+
+// Configs provides access to port configurations
+type Configs struct {
+	workspaceConfigs     map[uint32]*gitpod.PortConfig
+	instancePortConfigs  map[uint32]*gitpod.PortConfig
+	instanceRangeConfigs []*RangeConfig
+}
+
+// ForEach iterates over all configured ports
+func (configs *Configs) ForEach(callback func(port uint32, config *gitpod.PortConfig)) {
+	if configs == nil {
+		return
+	}
+	visited := make(map[uint32]struct{})
+	for _, configs := range []map[uint32]*gitpod.PortConfig{configs.instancePortConfigs, configs.workspaceConfigs} {
```

Comment:
Did not know you could join maps like that :)

Beware: maps do not guarantee an order, in fact they guarantee that the order is random.

Reply:
It should be fine here, order within the map is not important, only duplicates across.